### DISCUSSION TITLE
Is Cypress for unit tests or not?

### DISCUSSION THREAD

**Discussion by GaTechThomas** [cypress-io/cypress#16074] [General]
When I go to the intro "[how it works](https://www.cypress.io/how-it-works/)" page, the page says:

> Cypress is not a general automation framework, nor is it a unit testing framework for your back end services. There are already great tools out there that do that. Rather, we specialize in one thing - creating a great experience while you write end to end tests for your web applications.

But then the "[Who uses Cypress](https://docs.cypress.io/guides/overview/why-cypress#Who-uses-Cypress)" info says:

> Our users are typically developers or QA engineers building web applications using modern JavaScript frameworks.
> 
> Cypress enables you to write all types of tests:
> 
> - End-to-end tests
> - Integration tests
> - Unit tests
>
> Cypress can test anything that runs in a browser.
> 

Please help me to understand how both statements can be true.

**Comment by bahmutov** [comment]
Sure, a great question that has its own documentation issue here https://github.com/cypress-io/cypress-documentation/issues/1502

For unit tests, it depends _what_ the code does and what environment it expects to be able to execute. For example, you can test browser-specific code like `window.btoa` function. Cypress can execute the following unit test

```js
it('converts string to base64', () => {
  expect(window.btoa('foo')).to.equal('Zm9v')
})
```

You can also write a unit test for isomorphic or universal code that can run both in the browser and in Node. Cypress will happily run such tests

```js
import {add} from './my/math'
it('adds two numbers', () => {
  expect(add(2, 3)).to.equal(5)
})
```

But what Cypress _cannot_ do is to run unit tests for any code that needs Node environment - because Cypress executes the tests in the browser.

```js
const fs = require('fs')
it('reads a file', () => {
   // NOPE, NOT GOING TO WORK
  fs.readFileSync('file.txt', 'utf8')
})
```

This is why we say that Cypress can run all sorts of tests - as long as the code or the application is meant to run in the browser.

For more, see this presentation https://slides.com/bahmutov/no-excuses#/5 (click down to go through the slides) and https://slides.com/bahmutov/write-a-cypress-test#/only-tests (again, click the down arrow to go through the slides)

**Comment by GaTechThomas** [comment]
Thank you! Fantastic answer.
